### DISCUSSION TITLE
Cannot invoke "org.apache.rocketmq.store.timer.TimerMessageStore$TimerFlushService.shutdown()" because "this.timerFlushService" is null

### DISCUSSION THREAD

**Discussion by yidazhi** [apache/rocketmq#7245] [General]
mq版本：rocketmq-all-5.1.3-bin-release.zip
系统centos7
启动Broker报
java.lang.NullPointerException: Cannot invoke "org.apache.rocketmq.store.timer.TimerMessageStore$TimerFlushService.shutdown()" because "this.timerFlushService" is null
        at org.apache.rocketmq.store.timer.TimerMessageStore.shutdown(TimerMessageStore.java:511)
        at org.apache.rocketmq.broker.BrokerController.shutdownBasicService(BrokerController.java:1338)
        at org.apache.rocketmq.broker.BrokerController.shutdown(BrokerController.java:1464)
        at org.apache.rocketmq.broker.BrokerStartup.createBrokerController(BrokerStartup.java:242)
        at org.apache.rocketmq.broker.BrokerStartup.main(BrokerStartup.java:50)
请问如何处理

**Comment by joeCarf** [comment]
hi 贴一下你的指令 我看下能否复现

**Comment by RongtongJin** [comment]
> mq版本：rocketmq-all-5.1.3-bin-release.zip 系统centos7 启动Broker报 java.lang.NullPointerException: Cannot invoke "org.apache.rocketmq.store.timer.TimerMessageStore$TimerFlushService.shutdown()" because "this.timerFlushService" is null at org.apache.rocketmq.store.timer.TimerMessageStore.shutdown(TimerMessageStore.java:511) at org.apache.rocketmq.broker.BrokerController.shutdownBasicService(BrokerController.java:1338) at org.apache.rocketmq.broker.BrokerController.shutdown(BrokerController.java:1464) at org.apache.rocketmq.broker.BrokerStartup.createBrokerController(BrokerStartup.java:242) at org.apache.rocketmq.broker.BrokerStartup.main(BrokerStartup.java:50) 请问如何处理

这个应该不是broker启动不了的原因。标准输出有打印什么错误信息吗

**Comment by mxsm** [comment]
@RongtongJin @yidazhi  可以看一下这个 ISSUE #7269 .  导致的原因是可能配置的问题导致Broker的某些启动错误，具体是哪个配置的问题你可以贴出来你的配置或者自己对照官网排查一下。 然后触发退出的操作。在退出过程中 TimerMessageStore 的属性并没有实例化，在执行shutdown的时候空指针了。 @RongtongJin  麻烦review pr #7270 代码

**Comment by zswxjjh** [comment]
我也遇到了相关问题，MQ版本5.1.4，系统windwos11，使用./mqbroker -n 127.0.0.1:9876启动时出现一下报错
![屏幕截图 2023-11-28 091543](https://github.com/apache/rocketmq/assets/96217546/b94f1e85-95e5-4b55-9397-40b0ee151bdb)，
我在mqnamesrv.cmd和mqbroker.cmd中添加了ROCKETMQ_HOME变量
![屏幕截图 2023-11-28 091943](https://github.com/apache/rocketmq/assets/96217546/e3da1a0c-a145-4538-a7fd-0b782c51ce50)，
![屏幕截图 2023-11-28 092153](https://github.com/apache/rocketmq/assets/96217546/3ea79045-5402-4cd8-8dd2-dc98b94b4865)，
在runserver.cmd和runbroker.cmd中添加了JAVA_HOME变量，在runserver中我将
set "JAVA_OPT=%JAVA_OPT% -server -Xms2g -Xmx2g -Xmn1g -XX:MetaspaceSize=128m -XX:MaxMetaspaceSize=320m"修改为
set "JAVA_OPT=%JAVA_OPT% -server -Xms512m -Xmx1g -Xmn512m -XX:MetaspaceSize=128m -XX:MaxMetaspaceSize=320m"（因为我的内存实在没有那么大），在runbroker.cmd中将
set "JAVA_OPT=%JAVA_OPT% -server -Xms2g -Xmx2g"修改为
set "JAVA_OPT=%JAVA_OPT% -server -Xms512m -Xmx1g"，
并将
set "JAVA_OPT=%JAVA_OPT% -XX:MaxDirectMemorySize=15g"修改为
set "JAVA_OPT=%JAVA_OPT% -XX:MaxDirectMemorySize=2g"，
![屏幕截图 2023-11-28 092551](https://github.com/apache/rocketmq/assets/96217546/27d78b43-bb7e-4e77-b89a-8724f65180bd)，
![屏幕截图 2023-11-28 092645](https://github.com/apache/rocketmq/assets/96217546/c3c84382-9e48-48cc-a67f-c38acafaa1c3)，
有趣的是我在前一天运行一切正常，第二天打开电脑再次运行就出现了空指针异常，而我完全不知道原因

**Reply by joeCarf** [reply]
看起来和 @mxsm 大佬所说的一样，是由于启动失败导致的，这里的NPE应该不是失败的直接原因。如果你之前有过正常运行的情况，可以检查一下配置文件是否被更改或者之前的实例是否没有被正确关闭
